### PR TITLE
cnf-tests: networkAttachmentDefinitionBuilder refactoring

### PR DESCRIPTION
**Pull request by mmirecki**

### CODE REVIEW

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 1]
```diff
@@ -10,14 +10,37 @@ import (
...
+`
+
+	macVlanConfig = `"type": "macvlan"`
+
+	hostLocalIpam = `"type": "host-local", "subnet": "%s/24"`
+
+	staticIpam = `"type":"static","addresses":[{"address":"%s/24"}]`
+)
+
+type token struct {
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:31]: What's token? If I am getting it right, we always add the ipam to the last item.
Would it make sense to have ipam as a seperate builder field?

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 2]
```diff
@@ -31,42 +54,50 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
-        "ipam": {
-            "type": "host-local",
-            "subnet": "1.1.1.0/24"
-        }}`, bondName, link1, link2))
+	b.appendConfig(fmt.Sprintf(bondConfig, bondName, link1, link2))
+	return b
+}
+
+func (b *NetworkAttachmentDefinitionBuilder) WithStaticIpam(ip string) *NetworkAttachmentDefinitionBuilder {
+	b.tokens[len(b.tokens)-1].ipam = fmt.Sprintf(staticIpam, ip)
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:72]: Here we assume tokens has at least one item, which is not always true

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 1 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go]: Done

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 2 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go]: fixed

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 3]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
 func (b *NetworkAttachmentDefinitionBuilder) Build() (*netattdefv1.NetworkAttachmentDefinition, error) {
-	if len(b.configs) == 0 {
-		return nil, errors.New("NetworkAttachmentDefinition with no configs")
+	if b.errorMsg != "" {
+		return nil, errors.New(b.errorMsg)
 	}
-
-	b.definition.Spec.Config = fmt.Sprintf(`{"cniVersion":"0.4.0","name":"%s","plugins":[%s]}`, b.definition.ObjectMeta.Name, strings.Join(b.configs, ", "))
+	ipam := ""
+	if ipam != "" {
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:84]: b.ipam?

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 4]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
-	b.configs = append(b.configs, config)
+func (b *NetworkAttachmentDefinitionBuilder) WithTuning(sysctls string) *NetworkAttachmentDefinitionBuilder {
+	b.metaPluginConfigs = append(b.metaPluginConfigs, sysctls)
 	return b
 }
 
-func (b *NetworkAttachmentDefinitionBuilder) WithTuning(sysctls string) *NetworkAttachmentDefinitionBuilder {
-	b.configs = append(b.configs, sysctls)
+func (b *NetworkAttachmentDefinitionBuilder) WithBond(bondName, link1, link2 string) *NetworkAttachmentDefinitionBuilder {
+	b.setConfig(fmt.Sprintf(bondConfig, bondName, link1, link2))
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:60]: I'd leave the format string explicit here, so the behaviour of this function is clearer (and the const is not reused anywhere)

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 5]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
 	}
-
-	b.definition.Spec.Config = fmt.Sprintf(`{"cniVersion":"0.4.0","name":"%s","plugins":[%s]}`, b.definition.ObjectMeta.Name, strings.Join(b.configs, ", "))
+	ipam := ""
+	if ipam != "" {
+		ipam = fmt.Sprintf(`"ipam": {%s}, `, b.ipam)
+	}
+	configs := []string{fmt.Sprintf("{%s %s}", ipam, b.config)}
+	for _, metaPluginConfig := range b.metaPluginConfigs {
+		configs = append(configs, fmt.Sprintf("{%s %s}", ipam, metaPluginConfig))
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:89]: do we need ipam for each metaplugin?

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 6]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
-        }}`, bondName, link1, link2))
+func (b *NetworkAttachmentDefinitionBuilder) WithStaticIpam(ip string) *NetworkAttachmentDefinitionBuilder {
+	b.ipam = fmt.Sprintf(staticIpam, ip)
 	return b
 }
 
-func (b *NetworkAttachmentDefinitionBuilder) WithMacVlan(ip string) *NetworkAttachmentDefinitionBuilder {
-	b.configs = append(b.configs, fmt.Sprintf(`{"type": "macvlan","ipam":{"type":"static","addresses":[{"address":"%s/24"}]}}`, ip))
+func (b *NetworkAttachmentDefinitionBuilder) WithHostLocalIpam(ip string) *NetworkAttachmentDefinitionBuilder {
+	b.ipam = fmt.Sprintf(hostLocalIpam, ip)
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:70]: Same

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 7]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
-		"failOverMac": 1,
-		"linksInContainer": true,
-		"miimon": "100",
-		"links": [ {"name": "%s"}, {"name": "%s"} ],
-        "ipam": {
-            "type": "host-local",
-            "subnet": "1.1.1.0/24"
-        }}`, bondName, link1, link2))
+func (b *NetworkAttachmentDefinitionBuilder) WithStaticIpam(ip string) *NetworkAttachmentDefinitionBuilder {
+	b.ipam = fmt.Sprintf(staticIpam, ip)
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:65]: Same

File: cnf-tests/testsuites/pkg/networks/network.go [hunk 8]
```diff
@@ -30,43 +51,44 @@ func NewNetworkAttachmentDefinitionBuilder(namespace, nadName string) *NetworkAt
...
 
-func (b *NetworkAttachmentDefinitionBuilder) WithMacVlan(ip string) *NetworkAttachmentDefinitionBuilder {
-	b.configs = append(b.configs, fmt.Sprintf(`{"type": "macvlan","ipam":{"type":"static","addresses":[{"address":"%s/24"}]}}`, ip))
+func (b *NetworkAttachmentDefinitionBuilder) WithHostLocalIpam(ip string) *NetworkAttachmentDefinitionBuilder {
+	b.ipam = fmt.Sprintf(hostLocalIpam, ip)
+	return b
+}
+
+func (b *NetworkAttachmentDefinitionBuilder) WithMacVlan() *NetworkAttachmentDefinitionBuilder {
+	b.setConfig(macVlanConfig)
```
Comment (fedepaol) [inline cnf-tests/testsuites/pkg/networks/network.go:75]: Same

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 3 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go:84]: fxied

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 4 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go:60]: fixed

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 5 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go:89]: fixed

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 6 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go]: fixed

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 7 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go]: fixed

File: cnf-tests/testsuites/pkg/networks/network.go (hunk 8 above)
Comment (mmirecki) [reply cnf-tests/testsuites/pkg/networks/network.go]: fixed

**Pull request merged**